lib/web/addr.go

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package web
+
+import (
+	"bufio"
+	"net"
+	"net/http"
+	"net/netip"
+	"strings"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+const (
+	// xForwardedForHeader is a de-facto standard header for identifying the
+	// originating IP address of a client connecting to a web server through a
+	// proxy server.
+	xForwardedForHeader = "X-Forwarded-For"
+)
+
+// maybeUpdateClientSrcAddr overwrites client source address if X-Forwarded-For
+// is set.
+//
+// Both hijacked conn and request context are updated. The hijacked conn can be
+// used for ALPN connection upgrades or Websocket connections.
+func maybeUpdateClientSrcAddr(w http.ResponseWriter, r *http.Request) (http.ResponseWriter, *http.Request) {
+	// multiplexer handles extracting real client IP using PROXY protocol where
+	// available, so we can omit checking X-Forwarded-For.
+	if hasClientAddrFromMultiplexer(r.Context()) {
+		return w, r
+	}
+
+	// Use X-Forwarded-for if set.
+	xForwardedFor := r.Header.Get(xForwardedForHeader)
+	if xForwardedFor == "" {
+		return w, r
+	}
+
+	forwardedAddr, err := getForwardedAddr(r.RemoteAddr, xForwardedFor)
+	if err != nil {
+		logrus.Debugf("Invalid X-Forwarded-For %q: %v.", xForwardedFor, err)
+		return w, r
+	}
+
+	return responseWriterWithClientSrcAddr(w, forwardedAddr),
+		requestWithClientSrcAddr(r, forwardedAddr)
+}
+
+// getForwardedAddr returns a net.Addr from provided value of X-Forwarded-For.
+//
+// MDN reference:
+// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For
+//
+// AWS ALB reference:
+// https://docs.aws.amazon.com/elasticloadbalancing/latest/application/x-forwarded-headers.html
+func getForwardedAddr(observeredAddr, forwardedAddr string) (net.Addr, error) {
+	// In case multiple IPs are appended to X-Forwarded-For, use the first one.
+	forwardedAddr, _, _ = strings.Cut(forwardedAddr, ",")
```

Comment:
This assumption is not safe. Namely a client can add the  custom `X-Forwarding-For` header to the request and then forward this request through the proxy. In the result the AWS ALB will append  and update   the `X-Forwarded-For` header so received request on the Teleport Proxy side will have two values:
`X-Forwarding-For: ip-set-by-the-client-in-request, real-client-ip-set-by-the-aws-alb`  

where the first element can be set by the client.

Reply:
That's a good point. It makes this feature useless if a hacker can fake an ip. I will change it to use the last ip. Just in rare cases,  if there is more than one hop, the ip observed by the ALB may not be correct.